### PR TITLE
Prevent Pencil2D projects from referencing external files

### PR DESCRIPTION
**Pull request by scribblemaniac**
This PR adds additional checks to the loading of bitmap/vector/sound layers so that they will not be loaded if they are outside of the data directory. This can occur with specially crafted main.xml files, or with symlinks. This closes multiple admittedly unlikely security threats in the program that can result in exfiltrating images from arbitrary locations though animations, and even duplicating arbitrary files under some circumstances. Tests have also been written for the scenarios this PR covers with the exception of symlinks (and possibly Windows shortcuts) because Qt has very bad support for symlink creation.

This also incidentally fixes an issue where an infinite loop will occur when loading a main.xml file that contains a non-element node (ex. an xml comment) in the sound layer.

### CODE REVIEW

File: core_lib/src/structure/layerbitmap.cpp [hunk 1]
```diff
@@ -202,24 +202,41 @@ void LayerBitmap::loadDomElement(const QDomElement& element, QString dataDirPath
...
+
+            // Make sure src path is relative
+            shouldLoad &= QFileInfo(imageElement.attribute("src")).isRelative();
+            QFileInfo fi(dataDirPath, imageElement.attribute("src"));
+            // Make sure file is in data directory after resolving relative components and symlinks
+            QString canonicalPath = fi.canonicalFilePath();
+            fi.setFile(!canonicalPath.isEmpty() ? canonicalPath : fi.absoluteFilePath());
+            QDir dataDir(dataDirPath);
+            QDir ancestor = fi.dir();
+            while (ancestor != dataDir)
```
Comment (MrStevns) [inline core_lib/src/structure/layerbitmap.cpp:217]: Couldn't we move this to a method in Layer, so we don't have to have it duplicated three times?

File: core_lib/src/structure/layervector.cpp [hunk 2]
```diff
@@ -155,31 +155,53 @@ void LayerVector::loadDomElement(const QDomElement& element, QString dataDirPath
...
-                    getVectorImageAtFrame(position)->setOpacity(1.0);
-                progressStep();
             }
+            else
+            {
+                position = imageElement.attribute("frame").toInt();
+                addNewKeyFrameAt(position);
+                getVectorImageAtFrame(position)->loadDomElement(imageElement);
+                getVectorImageAtFrame(position)->setOpacity(imageElement.attribute("opacity", "1.0").toDouble());
+            }
```
Comment (MrStevns) [inline core_lib/src/structure/layervector.cpp:200]: I know that we had similar code but this feels wrong to me, why are we instantiating a keyframe if there's no src for it? Or let me rephrase, why do we create a new keyframe here at all when this method should only be about loading data if we have the necessary data.

It's not critique as much as a question, none of our other Layer classes have this so it seems weird to me that the Vector class has it, or do you disagree?

Comment (scribblemaniac) [reply core_lib/src/structure/layervector.cpp:200]: It is weird to me that they are handled differently between layers yes. I did not change this behavior because it's harmless and I wanted to minimize the changes I make so it can be merged quicker and with lower risk of introducing a bug in this critical file loading code.

I don't think it is wrong to instantiate an empty frame if there is no src. If there is a bitmap key with no data, it does not save an image for that frame, but it still sets the src. For a vector key a .vec file is written and src is set even if there is no data. An src attribute pointing to a file that does not exist and a file that contains no data both feel wrong to me. If there's no data that needs to be stored for a key, we shouldn't need to create a data file. If there's no data file, we should not need to specify the path to the non-existent file. But users still may want to save an emtpy key (to stop holding the previous key, or for planning/timing purposes).

File: core_lib/src/structure/layerbitmap.cpp (hunk 1 above)
Comment (scribblemaniac) [reply core_lib/src/structure/layerbitmap.cpp]: Yes very good point. I have refactored this into a separate function. It is in util because this is not exclusively used by layers now.

**Review by MrStevns — COMMENTED** [review]
I was about to approve this as it looks good to me now but i noticed that you haven't made this change for the camera layer, is that intentional?

**Comment by scribblemaniac** [comment]
> I was about to approve this as it looks good to me now but i noticed that you haven't made this change for the camera layer, is that intentional?

Yes this is intentional. As far as I know, the camera layer doesn't read any data from other files (all the camera data is the in the main.xml file) so it does not need these checks.

**Review by MrStevns — APPROVED** [review]
LGTM 👍

File: core_lib/src/util/util.cpp [hunk 3]
```diff
@@ -126,3 +128,59 @@ QString uniqueString(int len)
...
+    // Make sure src path is relative
+    if (!QFileInfo(filePath).isRelative()) return QString();
+
+    QFileInfo fi(dataDirPath, filePath);
+    // Recursively resolve symlinks
+    QString canonicalPath = fi.canonicalFilePath();
+
+    // Iterate over parent directories of the file path to see if one of them equals the data directory
+    QDir dataDir(dataDirPath);
+    if (dataDir.exists()) dataDir.setPath(dataDir.canonicalPath());
```
Comment (chchwy) [inline core_lib/src/util/util.cpp:143]: I don't really get what this line does? is it for resolving the symlinks?
Also could you add curly brackets?

File: core_lib/src/util/util.cpp [hunk 4]
```diff
@@ -126,3 +128,59 @@ QString uniqueString(int len)
...
+    if (canonicalPath.isEmpty())
+    {
+        // File does not exist, use absolute path and attempt to resolve symlinks again for each parent directory
+        fi.setFile(fi.absoluteFilePath());
+        QDir ancestor(fi.absoluteFilePath());
+        while (true) {
```
Comment (chchwy) [inline core_lib/src/util/util.cpp:149]: I don't quite understand this while loop. Could you please elaborate a bit more? probably with some examples?

File: core_lib/src/util/util.cpp (hunk 3 above)
Comment (scribblemaniac) [reply core_lib/src/util/util.cpp]: Yes. The canonical path is effectively the absolute path with all symlinks and relative components ('.' and '..') resolved. This particular line is necessary to deal with the case when one or more of the directories in the data dir's path is a symlink. For example, imagine if the data dir for a project is in `/tmp/Pencil2D/Project_Y2xD_012345/data`, and `/tmp` is actually a symlink to `/custom_temp_dir`. In this case, a vector file defined in main.xml could be `002.001.vec`, which when that file is fully resolved below, it will result in a path of `/custom_temp_dir/Pencil2D/Project_Y2xD_012345/data/002.001.vec`. This will not match the data dir value of `/tmp/Pencil2D/Project_Y2xD_012345/data` unless the symlinks are also resolved for the data dir path.

File: core_lib/src/util/util.cpp (hunk 4 above)
Comment (scribblemaniac) [reply core_lib/src/util/util.cpp:153]: For an example, let's say the data directory is at the path `/a/b/c` and the file path being validated is at `/a/b_symlink/c/d/file.txt`. The directory `b_symlink` is obviously a symlink to `/a/b`. We will also assume the directory `/a/b/c/d` and `file.txt` do not exist and directories `/a`, `/a/b`, and `/a/b/c` do exist. I will be walking through the logic based on the slightly rearranged code I have pushed since your review so pull the latest code before following along. 

This loop iterates up the directory hierarchy of the file path. Keep in mind that it only runs if the canonical path of the file is empty, which means the file and possibly some of its parent directories don't exist or are dangling symlinks (symlinks where the target location does not exist, QDir::exists() also returns false in this case).

Iteration 1:
1. ancestor is `/a/b_symlink/c/d/file.txt`
2. ancestor is not equal to data dir
3. newAncestor is set to the parent of ancestor: `/a/b_symlink/c/d`
4. newAncestor does not exist
5. newAncestor is different from ancestor
6. Set ancestor to newAncestor

Iteration 2:
1. ancestor is `/a/b_symlink/c/d`
2. ancestor is not equal to data dir
3. newAncestor is set to the parent of ancestor: `/a/b_symlink/c`
4. newAncestor does exist, so set new ancestor to the canonical path, which resolves the symlink b_symlink. newAncestor is now: `/a/b/c`
5. newAncestor is different from ancestor
6. Set ancestor to newAncestor

Iteration 3:
1. ancestor is `/a/b/c`
2. ancestor is equal to data dir, that means the file would be within the data directory after resolving all symlinks  if it and all of its parent directories existed. As a positive result, return the file path.

If the data dir was not `/a/b` and instead something like `/x`, then the loop would continue with ancestor = `/a/b` and then `/a` and then `/`. When ancestor is the root directory, calling dir on it will again return the root directory `/` and newAncestor will equal ancestor. As a negative result, return an invalid string because we now know that the file would not be within the data directory after resolving symlinks if it and all of its parent directories existed.

File: core_lib/src/util/util.cpp (hunk 3 above)
Comment (chchwy) [reply core_lib/src/util/util.cpp]: good thank you for explaining this.

File: core_lib/src/util/util.cpp (hunk 4 above)
Comment (chchwy) [reply core_lib/src/util/util.cpp:153]: Alright, thanks for the thorough explanation.

I now understand the nuance here. The condition `if (ancestor == newAncestor)` indicates that the path has reached the root directory, as the parent directory of the root is itself. This wasn’t immediately obvious to me, so I couldn't grasp it at first.

Would you be able to put this explanation into the code as a comment or a util function so the code can express its intent clearly? can `QDir::isRoot()` be used here?

**Review by chchwy — APPROVED** [review]

**Comment by scribblemaniac** [comment]
Thanks for the reviews @MrStevns and @chchwy

**Comment by chchwy** [comment]
All good. BTW, please use "Squash and Merge" next time. It gives a proper granularity of commits, making it clearer when reviewing the git history.
